src/vmspawn/vmspawn.c

```diff
@@ -0,0 +1,475 @@
+STATIC_DESTRUCTOR_REGISTER(arg_parameters, strv_freep);
+
+static int help(void) {
+        _cleanup_free_ char *link = NULL;
+        int r;
+
+        pager_open(arg_pager_flags);
+
+        r = terminal_urlify_man("systemd-vmspawn", "1", &link);
+        if (r < 0)
+                return log_oom();
+
+        printf("%1$s [OPTIONS...] [ARGUMENTS...]\n\n"
+               "%5$sSpawn a command or OS in a virtual machine.%6$s\n\n"
+               "  -h --help                 Show this help\n"
+               "     --version              Print version string\n"
+               "     --no-pager             Do not pipe output into a pager\n\n"
+               "%3$sImage:%4$s\n"
+               "  -i --image=PATH           Root file system disk image (or device node) for\n"
+               "                            the virtual machine\n"
+               "%3$sHost Configuration:%4$s\n"
+               "     --qemu-smp=SMP         Configure guest's SMP settings\n"
+               "     --qemu-mem=MEM         Configure guest's RAM size\n"
+               "     --qemu-kvm=auto|enabled|disabled\n"
+               "                            Configure whether to use KVM or not\n"
+               "     --qemu-gui             Start QEMU in graphical mode\n"
+               "%3$sCredentials:%4$s\n"
+               "     --set-credential=ID:VALUE\n"
+               "                            Pass a credential with literal value to container.\n"
+               "     --load-credential=ID:PATH\n"
+               "                            Load credential to pass to container from file or\n"
+               "                            AF_UNIX stream socket.\n"
+               "\nSee the %2$s for details.\n",
+               program_invocation_short_name,
+               link,
+               ansi_underline(),
+               ansi_normal(),
+               ansi_highlight(),
+               ansi_normal());
+
+        return 0;
+}
+
+static int parse_argv(int argc, char *argv[]) {
+        enum {
+                ARG_VERSION = 0x100,
+                ARG_NO_PAGER,
+                ARG_QEMU_SMP,
+                ARG_QEMU_MEM,
+                ARG_QEMU_KVM,
+                ARG_QEMU_GUI,
+                ARG_SET_CREDENTIAL,
+                ARG_LOAD_CREDENTIAL,
+        };
+
+        static const struct option options[] = {
+                { "help",            no_argument,       NULL, 'h'                 },
+                { "version",         no_argument,       NULL, ARG_VERSION         },
+                { "no-pager",        no_argument,       NULL, ARG_NO_PAGER        },
+                { "image",           required_argument, NULL, 'i'                 },
+                { "qemu-smp",        optional_argument, NULL, ARG_QEMU_SMP        },
+                { "qemu-mem",        optional_argument, NULL, ARG_QEMU_MEM        },
+                { "qemu-kvm",        optional_argument, NULL, ARG_QEMU_KVM        },
```

Comment:
`optional_argument` is something we try to avoid in new code, since it creates some weird ambiguities.

i.e. if you have a switch with an optional argument called `--foo`, what does an invocation like the following mean:

```c
myprogram --foo quux
```

is `quux` an argument to `--foo`, or is `--foo` specified without any argument, and `quux` is a positional argument to the program?

We ran into this a couple of times, and people are confused by this, hence we decided to avoid this, and only do `no_argument` and `required_argument` switches

Reply:
Thanks for the example showing the confusion.
My mental model of what required_argument vs optional argument was wrong while programming this.
I thought required_argument meant that providing that argument was required not that the the flag had to be given an argument.

Reply:
Fixed in 111681abaf.